RFC-0012-profile-directed-typing.md

```diff
@@ -0,0 +1,295 @@
+
+## PyAnnotate Profiler
+
+PyAnnotate (https://github.com/dropbox/pyannotate) is another open source tool that has similar functionality with MonkeyType. It also observes types of runtime values in Python execution and dumps function signatures. However, when compared to MonkeyType, it has following disadvantages:
+
+* PyAnnotate only generates Python2-style type lines, which requires additional parsing and may not be expressive enough in the future when Python evolves.
+* Unless we fork and modify their source code, PyAnnotate is less customizable. For example, it doesn’t support filtering functions/methods from other third party modules.
+* PyAnnotate is not as well-maintained, evident by the fact that it only supports Python2-style type line generation and that last meaningful commit into its repo is from April of 2020, while MonkeyType has commits almost every month.
+* PyAnnotate is owned by DropBox, while MonkeyType is maintained by Facebook/Instagram, therefore we are likely going to get better support.
+
+## Bidirectional Type Inference Algorithm
+
+This is another approach of solving the general problem of missing/incorrect type annotations. Basic idea is to devise an algorithm that infers type of an argument based on how it is used in function body. This approach carries a lot higher complexity and may not be a good fit for TorchScript language specifically.
+
+Stand-alone Tool with Human Intervention to Permanently Annotate Source Code
+
+Another option we thought about is to create a stand-alone tool that is based on MonkeyType to help users permanently annotate their source code, instead of only generating one-time type annotations on the fly as we proposed in this approach.
+
+The additional benefit is that stand-alone tool could optionally ask for user’s opinion on whether annotated types match their expectation and make revision based on user input.
+
+However, using a stand-alone tool has a few down-sides:
+
+* MonkeyType is unaware of TorchScript limitation of expressing nn.Module types and would blindly apply nn.Module and its subclasses as type annotation. Though legal in Python, they are not accepted in TorchScript. We need to configure MonkeyType to ensure it never annotates module types.
+* TorchScript has sophisticated type inference for return values of functions/methods, empirically it is more than enough to infer all return types of functions. Having MonkeyType annotate return types is more likely to cause type conflicts rather than helping with compilation.
+* User code may reference third-party libraries that user installed in system-wide locations. If we have a stand-alone tool that profiles types and applies derived type annotations to them, then user environment is polluted. If type annotations are not applied to them, then TorchScript would fail.
```

Comment:
```suggestion
* User code may reference third-party libraries that user has installed in system-wide locations. If we have a stand-alone tool that profiles types and applies derived type annotations, then the user environment would be polluted. On the other hand, if the type annotations were not applied, then TorchScript would fail.
```